### PR TITLE
[Api] adding product from another channel

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         |  1.9
| Bug fix?        | yes (kind of)
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | continues #12579 (#12604 now :D )
| License         | MIT

### CODE REVIEW

File: features/checkout/preventing_adding_to_cart_product_or_variant_from_another_channel.feature [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        And the store has a "Pain" product
+        And the store operates on a channel named "Poland" with hostname "poland"
+        And I am a logged in customer
+
+    @api
+    Scenario: Preventing customer from adding simple product from another channel
+        Given I am browsing channel "Poland"
+        When I pick up my cart
+        And I try to add product "Pain" to the cart
+        Then I should be informed that product "Pain" is not available in current channel
```
Comment (AdamKasp) [inline features/checkout/preventing_adding_to_cart_product_or_variant_from_another_channel.feature:20]:
```suggestion
        Then I should be informed that product "Pain" is not available
```
information about different channels shouldn't be provided to customer
now if you will add to cart a non-existent product you will get 500, so or we should have the same result here or validation message like "there is są product 'Pain' in shop'"

**Review by AdamKasp — CHANGES_REQUESTED** [review]

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantAvailableInChannelValidator.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        Assert::isInstanceOf($value, AddItemToCart::class);
+
+        /** @var ProductOrVariantAvailableInChannel $constraint */
+        Assert::isInstanceOf($constraint, ProductOrVariantAvailableInChannel::class);
+
+        /** @var OrderInterface $cart */
+        $cart = $this->orderRepository->findCartByTokenValue($value->getOrderTokenValue());
+
+        /** @var ProductVariantInterface $productVariant */
+        $productVariant = $this->productVariantRepository->findOneBy(['code' =>$value->productVariantCode]);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantAvailableInChannelValidator.php:54]:
```suggestion
        $productVariant = $this->productVariantRepository->findOneBy(['code' => $value->productVariantCode]);
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantAvailableInChannelValidator.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        Assert::notNull($cartChannel = $cart->getChannel());
+
+        /** @var ProductInterface $product */
+        $product = $productVariant->getProduct();
+
+        if(!$product->hasChannel($cartChannel)) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantAvailableInChannelValidator.php:61]:
```suggestion
        if (!$product->hasChannel($cartChannel)) {
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantAvailableInChannelValidator.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ProductOrVariantAvailableInChannelValidator extends ConstraintValidator
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantAvailableInChannelValidator.php:27]:
```suggestion
final class ProductAvailableInChannelValidator extends ConstraintValidator
```
?

File: src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml [hunk 5]
```diff
@@ -44,6 +44,8 @@ sylius:
         enabled:
             cannot_disable_base: The base locale cannot be disabled.
     product:
+        channel:
+            not_available: 'There is no product %productName% in this shop.'
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Resources/translations/validators.en.yml:48]: I would be for adding this message to the `validators.en.yaml` in `ApiBundle` as the validator is there
```suggestion
            not_available: 'The product %productName% does not exist.'
```

**Review by GSadee — COMMENTED** [review]
For me, preventing adding wrong product to the cart is a bug fix, so I would be for changing the base branch to at least `1.9`

**Comment by arti0090** [comment]
Closing as it is now moved to 1.9 on PR #12609

**Pull request closed without merging**